File: palette.go
// palette.go - Генератор цветовых палитр на Go (CLI)
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

// ========== ОСНОВНЫЕ ФУНКЦИИ ==========
type RGB struct {
	R int
	G int
	B int
}

type HSL struct {
	H float64
	S float64
	L float64
}

func hexToRgb(hex string) (RGB, error) {
	var r, g, b int
	hex = strings.TrimPrefix(hex, "#")
	if len(hex) == 3 {
		hex = string([]byte{hex[0], hex[0], hex[1], hex[1], hex[2], hex[2]})
	}
	if len(hex) != 6 {
		return RGB{}, fmt.Errorf("неверный HEX")
	}
	_, err := fmt.Sscanf(hex, "%02x%02x%02x", &r, &g, &b)
	return RGB{r, g, b}, err
}

func rgbToHex(r, g, b int) string {
	return fmt.Sprintf("#%02x%02x%02x", r, g, b)
}

func rgbToHsl(r, g, b int) HSL {
	rf := float64(r) / 255.0
	gf := float64(g) / 255.0
	bf := float64(b) / 255.0
	max := math.Max(rf, math.Max(gf, bf))
	min := math.Min(rf, math.Min(gf, bf))
	l := (max + min) / 2
	var h, s float64
	if max == min {
		h, s = 0, 0
	} else {
		d := max - min
		s = d / (1 - math.Abs(2*l-1))
		if max == rf {
			h = (gf - bf) / d
			if gf < bf {
				h += 6
			}
		} else if max == gf {
			h = (bf - rf) / d + 2
		} else {
			h = (rf - gf) / d + 4
		}
		h *= 60
		if h < 0 {
			h += 360
		}
	}
	return HSL{h, s * 100, l * 100}
}

func hslToRgb(h, s, l float64) RGB {
	h = math.Mod(h, 360)
	s /= 100
	l /= 100
	c := (1 - math.Abs(2*l-1)) * s
	x := c * (1 - math.Abs(math.Mod(h/60, 2)-1))
	m := l - c/2
	var r, g, b float64
	if h < 60 {
		r, g, b = c, x, 0
	} else if h < 120 {
		r, g, b = x, c, 0
	} else if h < 180 {
		r, g, b = 0, c, x
	} else if h < 240 {
		r, g, b = 0, x, c
	} else if h < 300 {
		r, g, b = x, 0, c
	} else {
		r, g, b = c, 0, x
	}
	return RGB{int((r + m) * 255), int((g + m) * 255), int((b + m) * 255)}
}

// ========== ГЕНЕРАЦИЯ ПАЛИТР ==========
func generatePalette(baseHex string, scheme string, count int) ([]RGB, error) {
	base, err := hexToRgb(baseHex)
	if err != nil {
		return nil, err
	}
	hsl := rgbToHsl(base.R, base.G, base.B)
	var colors []RGB
	switch scheme {
	case "mono":
		for i := 0; i < count; i++ {
			factor := float64(i) / float64(count-1)
			if count == 1 {
				factor = 0.5
			}
			newL := math.Max(10, math.Min(90, hsl.L*(0.5+factor)))
			colors = append(colors, hslToRgb(hsl.H, hsl.S, newL))
		}
	case "comp":
		colors = append(colors, base)
		colors = append(colors, hslToRgb(math.Mod(hsl.H+180, 360), hsl.S, hsl.L))
	case "analog":
		for i := 0; i < count; i++ {
			newH := math.Mod(hsl.H-30+float64(i)*(60/float64(count-1)), 360)
			colors = append(colors, hslToRgb(newH, hsl.S, hsl.L))
		}
	case "triad":
		for i := 0; i < 3; i++ {
			newH := math.Mod(hsl.H+float64(i)*120, 360)
			colors = append(colors, hslToRgb(newH, hsl.S, hsl.L))
		}
	case "tetrad":
		colors = append(colors, base)
		colors = append(colors, hslToRgb(math.Mod(hsl.H+180, 360), hsl.S, hsl.L))
		colors = append(colors, hslToRgb(math.Mod(hsl.H+90, 360), hsl.S, hsl.L))
		colors = append(colors, hslToRgb(math.Mod(hsl.H+270, 360), hsl.S, hsl.L))
	default:
		return nil, fmt.Errorf("неизвестная схема")
	}
	return colors, nil
}

// ========== ВИЗУАЛИЗАЦИЯ ==========
func printPalette(colors []RGB) {
	for i, c := range colors {
		hex := rgbToHex(c.R, c.G, c.B)
		bg := fmt.Sprintf("\033[48;2;%d;%d;%dm", c.R, c.G, c.B)
		reset := "\033[0m"
		lum := 0.299*float64(c.R) + 0.587*float64(c.G) + 0.114*float64(c.B)
		fg := "\033[38;2;255;255;255m"
		if lum < 140 {
			fg = "\033[38;2;255;255;255m"
		} else {
			fg = "\033[38;2;0;0;0m"
		}
		fmt.Printf("%s%s  %d. %s  RGB(%d,%d,%d)  %s\n", bg, fg, i+1, hex, c.R, c.G, c.B, reset)
	}
}

// ========== ЭКСПОРТ ==========
func exportCSS(colors []RGB, filename string) error {
	f, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer f.Close()
	f.WriteString(":root {\n")
	for i, c := range colors {
		f.WriteString(fmt.Sprintf("  --color-%d: %s;\n", i+1, rgbToHex(c.R, c.G, c.B)))
	}
	f.WriteString("}\n")
	return nil
}

func exportJSON(colors []RGB, filename string) error {
	type ColorData struct {
		R   int    `json:"r"`
		G   int    `json:"g"`
		B   int    `json:"b"`
		Hex string `json:"hex"`
	}
	data := make([]ColorData, len(colors))
	for i, c := range colors {
		data[i] = ColorData{c.R, c.G, c.B, rgbToHex(c.R, c.G, c.B)}
	}
	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, jsonData, 0644)
}

func exportHTML(colors []RGB, filename string) error {
	f, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer f.Close()
	f.WriteString("<html><head><style>body{font-family:sans-serif; display:flex; gap:20px;}</style></head><body>")
	for _, c := range colors {
		hex := rgbToHex(c.R, c.G, c.B)
		lum := 0.299*float64(c.R) + 0.587*float64(c.G) + 0.114*float64(c.B)
		color := "black"
		if lum < 140 {
			color = "white"
		}
		f.WriteString(fmt.Sprintf(`<div style="background:%s; width:100px; height:100px; border-radius:8px; text-align:center; line-height:100px; color:%s;">%s</div>`, hex, color, hex))
	}
	f.WriteString("</body></html>")
	return nil
}

// ========== ИСТОРИЯ ==========
type HistoryEntry struct {
	Date   string   `json:"date"`
	Scheme string   `json:"scheme"`
	Base   string   `json:"base"`
	Colors []string `json:"colors"`
}

const historyFile = "palette_history.json"

func loadHistory() []HistoryEntry {
	file, err := os.ReadFile(historyFile)
	if err != nil {
		return []HistoryEntry{}
	}
	var history []HistoryEntry
	json.Unmarshal(file, &history)
	return history
}

func saveHistoryEntry(entry HistoryEntry) {
	history := loadHistory()
	history = append(history, entry)
	if len(history) > 100 {
		history = history[len(history)-100:]
	}
	data, _ := json.MarshalIndent(history, "", "  ")
	os.WriteFile(historyFile, data, 0644)
}

// ========== CLI ==========
func main() {
	rand.Seed(time.Now().UnixNano())

	var (
		base     string
		scheme   string
		count    int
		random   bool
		export   string
		output   string
		history  bool
	)
	flag.StringVar(&base, "base", "", "Базовый цвет в HEX")
	flag.StringVar(&scheme, "scheme", "analog", "Схема (mono, comp, analog, triad, tetrad)")
	flag.IntVar(&count, "count", 5, "Количество цветов (2-10)")
	flag.BoolVar(&random, "random", false, "Случайный базовый цвет")
	flag.StringVar(&export, "export", "", "Экспорт (css, json, html)")
	flag.StringVar(&output, "output", "", "Имя файла для экспорта")
	flag.BoolVar(&history, "history", false, "Показать историю")
	flag.Parse()

	if history {
		hist := loadHistory()
		if len(hist) == 0 {
			fmt.Println("История пуста.")
		} else {
			fmt.Println("\n📋 ИСТОРИЯ ПАЛИТР")
			for _, entry := range hist {
				fmt.Printf("%s  %s  %d цветов\n", entry.Date[:19], entry.Scheme, len(entry.Colors))
			}
		}
		return
	}

	if random {
		r := rand.Intn(256)
		g := rand.Intn(256)
		b := rand.Intn(256)
		base = rgbToHex(r, g, b)
		fmt.Printf("Случайный базовый цвет: %s\n", base)
	} else if base == "" {
		reader := bufio.NewReader(os.Stdin)
		fmt.Print("Введите базовый цвет (HEX, например #FF5733): ")
		base, _ = reader.ReadString('\n')
		base = strings.TrimSpace(base)
		if !strings.HasPrefix(base, "#") {
			base = "#" + base
		}
	}

	colors, err := generatePalette(base, scheme, count)
	if err != nil {
		fmt.Println("Ошибка:", err)
		return
	}

	fmt.Printf("\n🎨 Палитра '%s' (база %s):\n", scheme, base)
	printPalette(colors)

	// Сохранить историю
	hexColors := make([]string, len(colors))
	for i, c := range colors {
		hexColors[i] = rgbToHex(c.R, c.G, c.B)
	}
	saveHistoryEntry(HistoryEntry{
		Date:   time.Now().Format(time.RFC3339),
		Scheme: scheme,
		Base:   base,
		Colors: hexColors,
	})

	if export != "" {
		filename := output
		if filename == "" {
			filename = fmt.Sprintf("palette.%s", export)
		}
		var err error
		switch export {
		case "css":
			err = exportCSS(colors, filename)
		case "json":
			err = exportJSON(colors, filename)
		case "html":
			err = exportHTML(colors, filename)
		default:
			fmt.Println("Неизвестный формат экспорта")
			return
		}
		if err != nil {
			fmt.Println("Ошибка экспорта:", err)
		} else {
			fmt.Printf("Экспортировано в %s\n", filename)
		}
	}
}

File: palette.rs
// palette.rs - Генератор цветовых палитр на Rust (CLI)
use clap::{Arg, App};
use rand::Rng;
use serde::{Serialize, Deserialize};
use std::fs;
use std::io::{self, Write};
use std::time::{SystemTime, UNIX_EPOCH};
use chrono::Local;

// ========== ОСНОВНЫЕ ФУНКЦИИ ==========
#[derive(Debug, Clone, Copy)]
struct RGB {
    r: u8,
    g: u8,
    b: u8,
}

struct HSL {
    h: f64,
    s: f64,
    l: f64,
}

fn hex_to_rgb(hex: &str) -> Result<RGB, String> {
    let hex = hex.trim_start_matches('#');
    if hex.len() == 3 {
        let chars: Vec<char> = hex.chars().collect();
        let expanded = format!("{}{}{}{}{}{}", chars[0], chars[0], chars[1], chars[1], chars[2], chars[2]);
        return hex_to_rgb(&expanded);
    }
    if hex.len() != 6 {
        return Err("Неверный HEX".to_string());
    }
    let r = u8::from_str_radix(&hex[0..2], 16).map_err(|_| "Ошибка парсинга")?;
    let g = u8::from_str_radix(&hex[2..4], 16).map_err(|_| "Ошибка парсинга")?;
    let b = u8::from_str_radix(&hex[4..6], 16).map_err(|_| "Ошибка парсинга")?;
    Ok(RGB { r, g, b })
}

fn rgb_to_hex(r: u8, g: u8, b: u8) -> String {
    format!("#{:02x}{:02x}{:02x}", r, g, b)
}

fn rgb_to_hsl(r: u8, g: u8, b: u8) -> HSL {
    let rf = r as f64 / 255.0;
    let gf = g as f64 / 255.0;
    let bf = b as f64 / 255.0;
    let max = rf.max(gf).max(bf);
    let min = rf.min(gf).min(bf);
    let l = (max + min) / 2.0;
    let (h, s) = if max == min {
        (0.0, 0.0)
    } else {
        let d = max - min;
        let s = d / (1.0 - (2.0 * l - 1.0).abs());
        let h = if max == rf {
            (gf - bf) / d + if gf < bf { 6.0 } else { 0.0 }
        } else if max == gf {
            (bf - rf) / d + 2.0
        } else {
            (rf - gf) / d + 4.0
        } * 60.0;
        (h % 360.0, s * 100.0)
    };
    HSL { h, s, l: l * 100.0 }
}

fn hsl_to_rgb(h: f64, s: f64, l: f64) -> RGB {
    let h = h % 360.0;
    let s = s / 100.0;
    let l = l / 100.0;
    let c = (1.0 - (2.0 * l - 1.0).abs()) * s;
    let x = c * (1.0 - ((h / 60.0) % 2.0 - 1.0).abs());
    let m = l - c / 2.0;
    let (r, g, b) = if h < 60.0 {
        (c, x, 0.0)
    } else if h < 120.0 {
        (x, c, 0.0)
    } else if h < 180.0 {
        (0.0, c, x)
    } else if h < 240.0 {
        (0.0, x, c)
    } else if h < 300.0 {
        (x, 0.0, c)
    } else {
        (c, 0.0, x)
    };
    RGB {
        r: ((r + m) * 255.0).round() as u8,
        g: ((g + m) * 255.0).round() as u8,
        b: ((b + m) * 255.0).round() as u8,
    }
}

// ========== ГЕНЕРАЦИЯ ПАЛИТР ==========
fn generate_palette(base_hex: &str, scheme: &str, count: usize) -> Result<Vec<RGB>, String> {
    let base = hex_to_rgb(base_hex)?;
    let hsl = rgb_to_hsl(base.r, base.g, base.b);
    let mut colors = Vec::new();
    match scheme {
        "mono" => {
            for i in 0..count {
                let factor = if count > 1 { i as f64 / (count - 1) as f64 } else { 0.5 };
                let new_l = (10.0f64).max((90.0f64).min(hsl.l * (0.5 + factor)));
                colors.push(hsl_to_rgb(hsl.h, hsl.s, new_l));
            }
        }
        "comp" => {
            colors.push(base);
            colors.push(hsl_to_rgb((hsl.h + 180.0) % 360.0, hsl.s, hsl.l));
        }
        "analog" => {
            for i in 0..count {
                let new_h = (hsl.h - 30.0 + i as f64 * (60.0 / (count - 1) as f64)) % 360.0;
                colors.push(hsl_to_rgb(new_h, hsl.s, hsl.l));
            }
        }
        "triad" => {
            for i in 0..3 {
                let new_h = (hsl.h + i as f64 * 120.0) % 360.0;
                colors.push(hsl_to_rgb(new_h, hsl.s, hsl.l));
            }
        }
        "tetrad" => {
            colors.push(base);
            colors.push(hsl_to_rgb((hsl.h + 180.0) % 360.0, hsl.s, hsl.l));
            colors.push(hsl_to_rgb((hsl.h + 90.0) % 360.0, hsl.s, hsl.l));
            colors.push(hsl_to_rgb((hsl.h + 270.0) % 360.0, hsl.s, hsl.l));
        }
        _ => return Err("Неизвестная схема".to_string()),
    }
    Ok(colors)
}

// ========== ВИЗУАЛИЗАЦИЯ ==========
fn print_palette(colors: &[RGB]) {
    for (i, c) in colors.iter().enumerate() {
        let hex = rgb_to_hex(c.r, c.g, c.b);
        let bg = format!("\x1b[48;2;{};{};{}m", c.r, c.g, c.b);
        let reset = "\x1b[0m";
        let lum = 0.299 * c.r as f64 + 0.587 * c.g as f64 + 0.114 * c.b as f64;
        let fg = if lum < 140.0 { "\x1b[38;2;255;255;255m" } else { "\x1b[38;2;0;0;0m" };
        println!("{}{}  {}. {}  RGB({},{},{})  {}", bg, fg, i+1, hex, c.r, c.g, c.b, reset);
    }
}

// ========== ЭКСПОРТ ==========
fn export_css(colors: &[RGB], filename: &str) -> Result<(), Box<dyn std::error::Error>> {
    let mut content = String::from(":root {\n");
    for (i, c) in colors.iter().enumerate() {
        content.push_str(&format!("  --color-{}: {};\n", i+1, rgb_to_hex(c.r, c.g, c.b)));
    }
    content.push_str("}\n");
    fs::write(filename, content)?;
    Ok(())
}

fn export_json(colors: &[RGB], filename: &str) -> Result<(), Box<dyn std::error::Error>> {
    #[derive(Serialize)]
    struct ColorData {
        r: u8,
        g: u8,
        b: u8,
        hex: String,
    }
    let data: Vec<ColorData> = colors.iter().map(|c| ColorData {
        r: c.r,
        g: c.g,
        b: c.b,
        hex: rgb_to_hex(c.r, c.g, c.b),
    }).collect();
    let json = serde_json::to_string_pretty(&data)?;
    fs::write(filename, json)?;
    Ok(())
}

fn export_html(colors: &[RGB], filename: &str) -> Result<(), Box<dyn std::error::Error>> {
    let mut content = String::from("<html><head><style>body{font-family:sans-serif; display:flex; gap:20px;}</style></head><body>");
    for c in colors {
        let hex = rgb_to_hex(c.r, c.g, c.b);
        let lum = 0.299 * c.r as f64 + 0.587 * c.g as f64 + 0.114 * c.b as f64;
        let text_color = if lum < 140.0 { "white" } else { "black" };
        content.push_str(&format!(
            r#"<div style="background:{}; width:100px; height:100px; border-radius:8px; text-align:center; line-height:100px; color:{};">{}</div>"#,
            hex, text_color, hex
        ));
    }
    content.push_str("</body></html>");
    fs::write(filename, content)?;
    Ok(())
}

// ========== ИСТОРИЯ ==========
#[derive(Serialize, Deserialize, Clone)]
struct HistoryEntry {
    date: String,
    scheme: String,
    base: String,
    colors: Vec<String>,
}

const HISTORY_FILE: &str = "palette_history.json";

fn load_history() -> Vec<HistoryEntry> {
    if let Ok(data) = fs::read_to_string(HISTORY_FILE) {
        if let Ok(entries) = serde_json::from_str(&data) {
            return entries;
        }
    }
    Vec::new()
}

fn save_history_entry(entry: HistoryEntry) {
    let mut history = load_history();
    history.push(entry);
    if history.len() > 100 {
        history = history[history.len()-100..].to_vec();
    }
    let json = serde_json::to_string_pretty(&history).unwrap();
    fs::write(HISTORY_FILE, json).unwrap();
}

// ========== CLI ==========
fn read_line(prompt: &str) -> String {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

fn main() {
    let matches = App::new("Palette Generator")
        .arg(Arg::with_name("base").long("base").takes_value(true).help("Базовый цвет в HEX"))
        .arg(Arg::with_name("scheme").long("scheme").takes_value(true).default_value("analog").help("Схема"))
        .arg(Arg::with_name("count").long("count").takes_value(true).default_value("5").help("Количество цветов"))
        .arg(Arg::with_name("random").long("random").help("Случайный базовый цвет"))
        .arg(Arg::with_name("export").long("export").takes_value(true).help("Экспорт (css, json, html)"))
        .arg(Arg::with_name("output").long("output").takes_value(true).help("Имя файла"))
        .arg(Arg::with_name("history").long("history").help("Показать историю"))
        .get_matches();

    if matches.is_present("history") {
        let history = load_history();
        if history.is_empty() {
            println!("История пуста.");
        } else {
            println!("\n📋 ИСТОРИЯ ПАЛИТР");
            for entry in history.iter().take(10) {
                println!("{}  {}  {} цветов", &entry.date[..19], entry.scheme, entry.colors.len());
            }
        }
        return;
    }

    let base = if matches.is_present("random") {
        let r = rand::thread_rng().gen_range(0..=255);
        let g = rand::thread_rng().gen_range(0..=255);
        let b = rand::thread_rng().gen_range(0..=255);
        let hex = rgb_to_hex(r, g, b);
        println!("Случайный базовый цвет: {}", hex);
        hex
    } else if let Some(b) = matches.value_of("base") {
        b.to_string()
    } else {
        read_line("Введите базовый цвет (HEX, например #FF5733): ")
    };

    let scheme = matches.value_of("scheme").unwrap_or("analog");
    let count: usize = matches.value_of("count").unwrap().parse().unwrap_or(5);

    match generate_palette(&base, scheme, count) {
        Ok(colors) => {
            println!("\n🎨 Палитра '{}' (база {}):", scheme, base);
            print_palette(&colors);
            // Сохранить историю
            let hex_colors: Vec<String> = colors.iter().map(|c| rgb_to_hex(c.r, c.g, c.b)).collect();
            save_history_entry(HistoryEntry {
                date: Local::now().to_rfc3339(),
                scheme: scheme.to_string(),
                base: base.clone(),
                colors: hex_colors,
            });
            // Экспорт
            if let Some(export_fmt) = matches.value_of("export") {
                let filename = matches.value_of("output").unwrap_or(&format!("palette.{}", export_fmt));
                let result = match export_fmt {
                    "css" => export_css(&colors, filename),
                    "json" => export_json(&colors, filename),
                    "html" => export_html(&colors, filename),
                    _ => Ok(()),
                };
                if let Err(e) = result {
                    println!("Ошибка экспорта: {}", e);
                } else {
                    println!("Экспортировано в {}", filename);
                }
            }
        }
        Err(e) => println!("Ошибка: {}", e),
    }
}
